segmentation/utils/spatial_analysis_utils.py

```diff
@@ -250,3 +250,57 @@ def calculate_enrichment_stats(close_num, close_num_rand):
     dims = ["stats", "rows", "cols"]
     stats_xr = xr.DataArray(stats_data, coords=coords, dims=dims)
     return stats_xr
+
+
+def compute_neighbor_counts(fov_data, dist_matrix, distlim,
+                            cell_neighbor_counts, cell_neighbor_freqs,
+                            cell_label_col="cellLabelInImage", cell_type_col="cell_type"):
+    """Calculates the number of neighbor phenotypes for each cell. The cell counts itself as a neighbor.
+
+    Args:
+        fov_data: data for the current fov
+        dist_matrix: cells x cells matrix with the euclidian
+            distance between centers of corresponding cells
+        distlim: threshold for spatial enrichment distance proximity
+        cell_neighbor_counts: matrix with phenotype counts per cell
+        cell_neighbor_freqs: matrix with phenotype frequencies of
+            counts per phenotype/total phenotypes for each cell
+        cell_label_col: Column name with the cell labels
+        cell_type_col: column name with the cell phenotypes
+    Returns:
+        cell_neighbor_counts: matrix with phenotype counts per cell
+        cell_neighbor_freqs: matrix with phenotype frequencies of
+            counts per phenotype/total phenotypes for each cell
+        cell_count: current cell in analysis"""
+
+    # Get the phenotypes included in the current fov
+    pheno_titles = fov_data[cell_type_col].drop_duplicates()
+
+    # remove non-cell2cell distances
+    cell_dist_mat = np.take(dist_matrix, fov_data[cell_label_col] - 1, 0)
+    cell_dist_mat = np.take(cell_dist_mat, fov_data[cell_label_col] - 1, 1)
+
+    # binarize distance matrix
+    cell_dist_mat_bin = np.zeros(cell_dist_mat.shape)
+    cell_dist_mat_bin[cell_dist_mat < distlim] = 1
+    # cell_dist_mat_bin[cell_dist_mat == 0] = 0 <- cell counts itself as neighbor if this line is not included
+
+    # get num_neighbors for freqs
+    num_neighbors = np.sum(cell_dist_mat_bin, axis=0)
+
+    # create the 'phenotype has cell?' matrix, excluding non cell-label rows
+    pheno_has_cell = pd.get_dummies(fov_data.iloc[:, 2]).to_numpy().T
+
+    # dot binarized 'is neighbor?' matrix with pheno_has_cell to get counts
+    counts = pheno_has_cell.dot(cell_dist_mat_bin).T
+
+    # compute freqs with num_neighbors
+    freqs = counts.T / num_neighbors
+
+    # convert to pandas df and assign columns to phenos
+    countsdf = pd.DataFrame(counts, columns=pheno_titles)
+    freqsdf = pd.DataFrame(freqs.T, columns=pheno_titles)
+
+    # add to neighbor counts + freqs for only the matching phenotypes between the fov and the whole dataset
+    cell_neighbor_counts.loc[fov_data.index, countsdf.columns] = counts
+    cell_neighbor_freqs.loc[fov_data.index, freqsdf.columns] = freqs.T
```

Comment:
I think we can bypass making new df's by indexing with `.loc[fov_data.index, pheno_titles]` directly.  I could be wrong tho, I don't work with df's too often.